File: BGOI/bgoi09p4/src/bgoi09p4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class bgoi09p4 {
    /* 
     * Copy-pasting code is NOT cool! Please do not copy and paste my code as a submission to DMOJ.
     * github.com/jdabtieu/competitive-programming
     */
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer in;

    public static void main(String[] args) throws IOException {
        int n = readInt();
        int[] heights = new int[n];
        int[] ans = new int[n];
        for (int i = n - 1; i >= 0; i--) {
            heights[i] = readInt();
        }
        for (int i = 0; i < n; i++) {
            ans[i] = readInt();
        }
        int[] stack = new int[n];
        int ptr = -1;
        for (int i = 0; i < n; i++) {
            while (ptr >= 0 && stack[ptr] <= heights[i]) ptr--;
            
            if (ans[n-i-1] == 0) {
                ans[n-i-1] = heights[i];
            } else if (ans[n-i-1] <= (ptr+1)) {
                ans[n-i-1] = stack[ptr+1-ans[n-i-1]];
            } else {
                ans[n-i-1] = -1;
            }
            
            stack[++ptr] = heights[i];
        }
        for (int a : ans) System.out.print(a + " ");
    }

    static String next() throws IOException {
        while (in == null || !in.hasMoreTokens())
            in = new StringTokenizer(br.readLine());
        return in.nextToken();
    }

    static long readLong() throws IOException {
        return Long.parseLong(next());
    }
    
    static int readInt() throws IOException{
        int x = 0, c;
        while((c = br.read()) != ' ' && c != '\n')
            x = x * 10 + (c - '0');
        return x;
    }

    static double readDouble() throws IOException {
        return Double.parseDouble(next());
    }

    static char readChar() throws IOException {
        return next().charAt(0);
    }

    static String readLine() throws IOException {
        return br.readLine();
    }
}

File: BGOI/bgoi09p4/src/bgoi09p4.cpp
#include <unistd.h>
#include <sys/syscall.h>
#pragma GCC optimize ("Ofast")
#define pb push_back
typedef unsigned long long llu;
typedef long long ll;
typedef unsigned int uint;
typedef __uint128_t lllu;
typedef __int128_t lll;

int heights[1000001];
int ans[1000001];
int stk[1000001];

#define BUF_SIZE 65536

char ibuf[BUF_SIZE];
char obuf[BUF_SIZE];
int iPtr = 0;
int maxPtr = 0;
int oPtr = 0;

void Read() {
    maxPtr = syscall(SYS_read, 0, &ibuf, BUF_SIZE);
    iPtr = 0;
}

void Write() {
    syscall(SYS_write, 1, &obuf, oPtr);
    oPtr = 0;
}

void pc(char c) {
    if (oPtr >= BUF_SIZE) Write();
    obuf[oPtr++] = c;
}

void _pi(int n) {
    if (n == 0) return;
    _pi(n / 10);
    pc((char) (n % 10 + '0'));
}

void pi(int num) {
    if (num < 0) {
        pc('-');
        _pi(-num);
    } else if (num == 0) {
        pc('0');
    } else {
        _pi(num);
    }
}

char sc() {
    if (iPtr >= maxPtr) Read();
    return ibuf[iPtr++];
}

int si() {
    int x = 0;
    char c;
    while ((c = sc()) <= ' ');
    int neg = 0;
    if (c == '-') {
        neg = 1;
        c = sc();
    }
    do {
        x = x * 10 + (c - '0');
    } while ((c = sc()) > ' ');
    return neg ? -x : x;
}

unsigned int su() {
    int x = 0;
    char c;
    while ((c = sc()) <= ' ');
    do {
        x = x * 10 + (c - '0');
    } while ((c = sc()) > ' ');
    return x;
}
/*
 * Copy-pasting code is NOT cool! Please do not copy and paste my code as a submission to DMOJ.
 * github.com/jdabtieu/competitive-programming
 */
int main() {
        int n = su();
        for (int i = n - 1; i >= 0; i--) {
            heights[i] = su();
        }
        for (int i = 0; i < n; i++) {
            ans[i] = su();
        }
        int ptr = -1;
        for (int i = 0; i < n; i++) {
            while (ptr >= 0 && stk[ptr] <= heights[i]) ptr--;

            if (ans[n-i-1] == 0) {
                ans[n-i-1] = heights[i];
            } else if (ans[n-i-1] <= (ptr+1)) {
                ans[n-i-1] = stk[ptr+1-ans[n-i-1]];
            } else {
                ans[n-i-1] = -1;
            }

            stk[++ptr] = heights[i];
        }
        for (int i = 0; i < n; i++) {
            pi(ans[i]);
            pc(' ');
        }
        Write();
    }
